# About All by All

The All by All browser maps known and novel associations between genotypes and phenotypes using data contributed by *All of Us* Research Program participants as of July 1, 2022\. All by All encompasses about 3,400 phenotypes with gene-based and single-variant associations across nearly 250,000 whole genome sequences. In total, about 500 billion associations are available to explore. Results for individual common variants and group tests for rare variants are available, such as burden tests of predicted loss-of-function (pLoF) against each phenotype, in a public, cloud-based browser.

The All by All analyses use data from the Curated Data Repository (CDR) v7 data release (April 2023). The results will be updated periodically as more genomic data are available. The data for the phenotypes were drawn from [disease phenotypes derived from electronic health record data](https://www.researchallofus.org/data-tools/data-sources/?utm_source=nih&utm_medium=referral&utm_campaign=All-by-All-Browser&utm_term=About-text&utm_content=Data-Sources) on medication exposures, conditions, and lab measurements; participants’ [physical measurements](https://www.researchallofus.org/data-tools/data-sources/?utm_source=nih&utm_medium=referral&utm_campaign=All-by-All-Browser&utm_term=About-text&utm_content=Data-Sources); and participants’ responses to the [Personal and Family Health History survey](https://www.researchallofus.org/data-tools/survey-explorer/?utm_source=nih&utm_medium=referral&utm_campaign=All-by-All-Browser&utm_term=About-text&utm_content=Survey-Explorer). Analyses were run separately by computed genetic ancestry groups and then combined into a meta-analysis. [Read more about how *All of Us* computes genetic ancestry](https://aousupporthelp.zendesk.com/hc/en-us/articles/4614687617556#h_01GY7QYC017WWMXFB0YTT233G1). 

 With this publicly accessible browser, anyone can explore known or novel associations without prior genomic research or coding experience. To further explore the All by All tables, and use it to jumpstart workspace analysis, researchers can [log in to or register for the Researcher Workbench](https://www.researchallofus.org/register/?utm_source=nih&utm_medium=referral&utm_campaign=All-by-All-Browser&utm_term=About-text&utm_content=Register).

Within that platform, researchers can combine the All by All tables with additional data types to find more details about novel associations, dig deeper into understudied conditions, study potential drug targets, or validate findings from other studies. Note: The All by All browser was built using Hail Tables and SQLite. When comparing results between the All by All browser and the All by All tables in the Researcher Workbench, researchers should use the Hail Tables available on the Researcher Workbench, rather than the Matrix Tables.

In accordance with program policy, the All by All Browser applied for an exception to the [Data and Statistics Dissemination Policy](https://www.researchallofus.org/faq/data-and-statistics-dissemination-policy/) to display genotype-phenotype associations with participant counts fewer than 20. The program granted this exception on May 17, 2024, in light of the browser’s scientific utility and minimal risk to participant privacy. For more information about program policies and policy compliance, please visit the Data Access Tiers page on the <em>All of Us</em> Research Hub.

Additional All by All resources and support materials will be released in the coming months. These will include additional information about the context, limitations, and appropriate use of this resource.

Questions? Contact [support@researchallofus.org](mailto:support@researchallofus.org).
